src/core/document/json.pl

```diff
@@ -368,39 +368,54 @@
     path_strings(Rev, Prefixes, Strings),
     merge_separator_split(Path_String, '/', Strings).
 
-json_idgen(JSON,DB,Context,Path,ID_Ex) :-
-    get_dict('@type',JSON,Type),
-    key_descriptor(DB,Context,Type,Descriptor),
-    (   Descriptor = lexical(Base,Fields)
-    ->  get_field_values(JSON, DB, Context, Fields, Values),
-        path_component([type(Base)|Path], Context, [Path_Base]),
-        idgen_lexical(Path_Base,Values,ID)
-    ;   Descriptor = hash(Base,Fields)
-    ->  get_field_values(JSON, DB, Context, Fields, Values),
-        path_component([type(Base)|Path], Context, [Path_Base]),
-        idgen_hash(Path_Base,Values,ID)
-    ;   Descriptor = value_hash(Base)
-    ->  get_all_path_values(JSON,Path_Values),
-        idgen_path_values_hash(Base,Path_Values,ID)
-    ;   (   Descriptor = random(Base)
-        ;   Descriptor = base(Base))
-    ->  (   get_dict('@id', JSON, Submitted_ID),
-            ground(Submitted_ID)
-        ->  path_component([type(Base)|Path], Context, [Path_Base]),
-            idgen_check_base(Submitted_ID, Path_Base, Context),
-            ID = Submitted_ID
-        ;   path_component([type(Base)|Path], Context, [Path_Base]),
-            idgen_random(Path_Base,ID)
-        )
+json_idgen(lexical(Base, Fields), JSON, DB, Context, Path, Id) :-
+    !,
+    get_field_values(JSON, DB, Context, Fields, Values),
+    path_component([type(Base)|Path], Context, [Path_Base]),
+    idgen_lexical(Path_Base, Values, Id).
+json_idgen(hash(Base, Fields), JSON, DB, Context, Path, Id) :-
+    !,
+    get_field_values(JSON, DB, Context, Fields, Values),
+    path_component([type(Base)|Path], Context, [Path_Base]),
+    idgen_hash(Path_Base, Values, Id).
+json_idgen(value_hash(Base), JSON, _DB, _Context, _Path, Id) :-
+    !,
+    get_all_path_values(JSON, Path_Values),
+    idgen_path_values_hash(Base, Path_Values, Id).
+json_idgen(Descriptor, JSON, _DB, Context, Path, Id) :-
+    (   Descriptor = random(Base)
+    ;   Descriptor = base(Base)
```

Comment:
This is better than before! However, stylistically, it's slightly better to have no explicit cuts, and to instead use first argument indexing, which means no fall-through case (two clauses for `random` and `base`).

Reply:
Is it better now?

Reply:
Yup!